firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java

```diff
@@ -269,16 +353,68 @@ void recalculateAndSaveOverlays(Set<DocumentKey> documentKeys) {
     // Apply the overlays and match against the query.
     ImmutableSortedMap<DocumentKey, Document> results = emptyDocumentMap();
     for (Map.Entry<DocumentKey, MutableDocument> docEntry : remoteDocuments.entrySet()) {
-      Mutation overlay = overlays.get(docEntry.getKey());
-      if (overlay != null) {
-        overlay.applyToLocalView(docEntry.getValue(), null, Timestamp.now());
+      Overlay overlay = overlays.get(docEntry.getKey());
+      Mutation mutation = overlay != null ? overlay.getMutation() : null;
+      if (mutation != null) {
+        mutation.applyToLocalView(
+            docEntry.getValue(), null, overlay.getLargestBatchId(), Timestamp.now());
       }
       // Finally, insert the documents that still match the query
       if (query.matches(docEntry.getValue())) {
         results = results.insert(docEntry.getKey(), docEntry.getValue());
       }
     }
+    return results;
+  }
+
+  /**
+   * Returns a mapping of documents to their largest batch ids based on the earliest batch id in the
+   * provided offset.
+   */
+  public Map<Document, Integer> getDocumentsInOverlayCacheByBatchId(
+      Query query, IndexOffset offset) {
+    Map<DocumentKey, MutableDocument> remoteDocuments = new HashMap<>();
+    Map<DocumentKey, Overlay> overlays =
+        documentOverlayCache.getOverlays(query.getPath(), offset.getLargestBatchId());
+
+    // Apply overlays to empty mutable document map to generate base.
+    for (Map.Entry<DocumentKey, Overlay> entry : overlays.entrySet()) {
+      remoteDocuments.put(entry.getKey(), MutableDocument.newInvalidDocument(entry.getKey()));
+    }
+
+    // Apply the overlays and match against the query.
+    Map<Document, Integer> results = new HashMap<>();
+    for (Map.Entry<DocumentKey, MutableDocument> docEntry : remoteDocuments.entrySet()) {
+      Overlay overlay = overlays.get(docEntry.getKey());
+      int batchId = overlay.getLargestBatchId();
+      Mutation mutation = overlay.getMutation();
+      if (mutation != null) {
+        mutation.applyToLocalView(docEntry.getValue(), null, batchId, Timestamp.now());
+      }
+      // Finally, insert the documents that still match the query.
+      if (query.matches(docEntry.getValue())) {
```

Comment:
I think we should do this once at the end of `getNextDocumentsAndOffset`.

FWIW, we might not need to do this at all if we pass a collection group here instead of the query. None of the IndexBackfill methods need a query.